2015-Fall/lrusnac/CalculatorV2/CalculatorV2/CalculatorViewController.swift

```diff
@@ -0,0 +1,119 @@
+//
+//  CalculatorViewController.swift
+//  CalculatorV2
+//
+//  Created by Leonid Rusnac on 08/09/15.
+//  Copyright (c) 2015 Leonid Rusnac. All rights reserved.
+//
+
+import UIKit
+
+class CalculatorViewController: UIViewController {
+
+    @IBOutlet weak var display: UILabel!
+    
+    var calculatorModel = CalculatorModel()
+    var userIsInTheMiddleOfTypingANumber = false
+    var pointWasInsertedInThisNumber = false
+    
+    @IBAction func appendDigit(sender: UIButton) {
+        if let digit = sender.currentTitle {
+            if userIsInTheMiddleOfTypingANumber {
+                display.text = display.text! + digit
+            } else {
+                display.text = digit
+                userIsInTheMiddleOfTypingANumber = true
+            }
+        }
+    }
+
+    @IBAction func clear() {
+        calculatorModel.clear()
+        display.text = "\(0)"
+        userIsInTheMiddleOfTypingANumber = false
+        pointWasInsertedInThisNumber = false
+    }
+    
+    @IBAction func backspace() {
+        // I let the user to backspace only if it is in middle of typing
+        if userIsInTheMiddleOfTypingANumber {
+            if let displayText = display.text {
+                let newDisplay = dropLast(displayText)
```

Comment:
`userIsInTheMiddleOfTypingANumber`is just a boolean that you set to false and true as you go. If this for some reason gets out of sync with the reality of the program you want to fail gracefully, and not with your application crashing. Your logic is sound, but the resulting program is very brittle. You program with the assumption that `userIsInTheMiddleOfTypingANumber == true` implies that there is text in the display. You're basing the logic of your program on the semantics and assumption you have. This is problematic because if these semantics change your program might fail in ways you don't expect, and it makes your program harder to understand for outsiders.

You check that `display.text` has a value (with the nice if-let around it), why not be sure about content as well?

To summerise, what you're doing is not wrong, but it is something that can potentially cause problems down the line.